File: binding-csharp/runtime/src/main/csharp/Etch/Util/SingleIterator.cs
// $Id$
// 
// Copyright 2007-2008 Cisco Systems Inc.
// 
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy
// of the License at
//  
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations
// under the License.

using System;
using System.Collections.Generic;

namespace Etch.Util
{
    /// <summary>
    /// An iterator over a single object
    /// </summary>
    public class SingleIterator<E> : IEnumerator<E>
    {
        private E _obj;

        public SingleIterator( Object obj )
        {
            _obj = ( E ) obj;
        }

        #region IEnumerator<E> Members

        public E Current
        {
            get 
            {
                if ( _obj == null )
                    throw new Exception( "No Such element" );

                E o = _obj;
                _obj = default(E);
                return o;
            }
        }

        #endregion

        #region IDisposable Members

        public void Dispose()
        {
            throw new Exception( "The method or operation is not implemented." );
        }

        #endregion

        #region IEnumerator Members

        object System.Collections.IEnumerator.Current
        {
            get
            {
                if ( _obj == null )
                    throw new Exception( "No Such element" );

                E o = _obj;
                _obj = default( E );
                return o;
            }
        }

        public bool MoveNext()
        {
            return _obj != null;
        }

        public void Reset()
        {
            throw new Exception( "The method or operation is not implemented." );
        }

        #endregion
    }
}

File: util/src/main/java/org/apache/etch/util/SingleIterator.java
/* $Id$
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.etch.util;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * An iterator over a single object.
 * @param <E>
 */
final public class SingleIterator<E> implements Iterator<E>
{
	/**
	 * Constructs a SingleIterator.
	 *
	 * @param obj the object to be iterated.
	 */
	public SingleIterator( E obj )
	{
		this.obj = obj;
	}
	
	private E obj;
	
	/* (non-Javadoc)
	 * @see java.util.Iterator#hasNext()
	 */
	public boolean hasNext()
	{
		return obj != null;
	}

	/* (non-Javadoc)
	 * @see java.util.Iterator#next()
	 */
	public E next()
	{
		if (obj == null)
			throw new NoSuchElementException();
		
		E o = obj;
		obj = null;
		return o;
	}

	/* (non-Javadoc)
	 * @see java.util.Iterator#remove()
	 */
	public void remove()
	{
		throw new UnsupportedOperationException();
	}
}
